source/common/buffer/buffer_impl.cc

```diff
@@ -217,21 +638,69 @@ OwnedImpl::OwnedImpl(const Instance& data) : OwnedImpl() { add(data); }
 OwnedImpl::OwnedImpl(const void* data, uint64_t size) : OwnedImpl() { add(data, size); }
 
 std::string OwnedImpl::toString() const {
-  uint64_t num_slices = getRawSlices(nullptr, 0);
-  STACK_ARRAY(slices, RawSlice, num_slices);
-  getRawSlices(slices.begin(), num_slices);
-  size_t len = 0;
-  for (const RawSlice& slice : slices) {
-    len += slice.len_;
+  if (old_impl_) {
+    uint64_t num_slices = getRawSlices(nullptr, 0);
+    STACK_ARRAY(slices, RawSlice, num_slices);
+    getRawSlices(slices.begin(), num_slices);
+    size_t len = 0;
+    for (const RawSlice& slice : slices) {
+      len += slice.len_;
+    }
+    std::string output;
+    output.reserve(len);
+    for (const RawSlice& slice : slices) {
+      output.append(static_cast<const char*>(slice.mem_), slice.len_);
+    }
+
+    return output;
+  } else {
+    uint64_t num_slices = getRawSlices(nullptr, 0);
+    if (num_slices == 0) {
+      return std::string();
+    }
+    STACK_ARRAY(slices, RawSlice, num_slices);
+    getRawSlices(slices.begin(), num_slices);
+    size_t len = 0;
+    for (const RawSlice& slice : slices) {
+      len += slice.len_;
+    }
+    std::string output;
+    output.reserve(len);
+    for (const RawSlice& slice : slices) {
+      output.append(static_cast<const char*>(slice.mem_), slice.len_);
+    }
+
+    return output;
   }
-  std::string output;
-  output.reserve(len);
-  for (const RawSlice& slice : slices) {
-    output.append(static_cast<const char*>(slice.mem_), slice.len_);
+}
+
+void OwnedImpl::postProcess() {}
+
+void OwnedImpl::appendSliceForTest(const void* data, uint64_t size) {
+  if (old_impl_) {
+    OwnedImpl rhs(data, size);
+    move(rhs);
+  } else {
+    slices_.emplace_back(OwnedSlice::create(data, size));
+    length_ += size;
   }
+}
 
-  return output;
+void OwnedImpl::appendSliceForTest(absl::string_view data) {
+  appendSliceForTest(data.data(), data.size());
 }
 
+void OwnedImpl::useOldImpl(bool use_old_impl) { use_old_impl_ = use_old_impl; }
```

Comment:
To ensure that this only gets called during the quiescent state between tests, there's a simple way now.

Have a static bool committed_ alongside use_old_impl_. In this method do:
```
void OwnedImpl::useOldImpl(bool use_old_impl) {
  ASSERT(!committed_);
  use_old_impl_ = use_old_impl;
}
```

then you can write:
```
void OwnedImpl::clearCommitted() { committed_ = false; }
```
which you can call from here:

https://github.com/envoyproxy/envoy/blob/741df7aa13ddb15acbc77468579f0c6b58542839/test/test_listener.cc#L15

That will allow switching implementations between test methods.

Actually I'm surprised this didn't bite you before, aren't you using TEST_P to test with both mechanisms in the same binary?

Reply:
Up until [cfd9d4a](https://github.com/envoyproxy/envoy/commit/cfd9d4a068e30f8fa29e7474496fe2e493b8cd01), the buffer unit tests used a separate, test-only method that enabled the implementation to be changed.

Starting with that change, it's valid to switch back and forth between the implementations in the same process. The one remaining restriction is that two buffers can only interact (e.g., with `buffer1.move(buffer2)`) if they use the same implementation.

That enables the coverage test to progress further, but eventually it still hangs with an exception somewhere. It will take some more digging to find out where that's happening, as the hang happens in some test code that has caught the exception and is trying to log it.

Reply:
Right; what I'm trying to suggest is that it still be invalid to switch between implementations within a process, allowing only for switching after a test method.

That should cover the case of the coverage tests, and also fuzz tests, both of which will work with the test listener as described.

I think we probably want to RELEASE_ASSERT, actually, that we only call useOldImpl once per process/test-method.

My thought was that like catching things at this level would be a stronger guarantee of the invariant that only one impl be alive at a time.

Of course you still have to finish your debugging :) There's a chance this suggestion may catch it, but it's probably something else.